transaction-multi-payment/src/lib.rs

```diff
@@ -361,68 +386,161 @@ where
+                    .checked_mul_int(fee)
+                    .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                match MC::withdraw(currency.into(), who, converted_fee) {
+                    Ok(()) => Ok(Some(PaymentInfo::NonNative(converted_fee, currency, price))),
+                    Err(_) => Err(InvalidTransaction::Payment.into()),
+                }
+            }
+        }
+    }
+
+    /// Since the predicted fee might have been too high, parts of the fee may
+    /// be refunded.
+    ///
+    /// Note: The `fee` already includes the `tip`.
+    fn correct_and_deposit_fee(
+        who: &T::AccountId,
+        _dispatch_info: &DispatchInfoOf<T::Call>,
+        _post_info: &PostDispatchInfoOf<T::Call>,
+        corrected_fee: Self::Balance,
+        tip: Self::Balance,
+        already_withdrawn: Self::LiquidityInfo,
+    ) -> Result<(), TransactionValidityError> {
+        let fee_receiver = DP::get_fee_receiver();
+
+        if let Some(paid) = already_withdrawn {
+            // Calculate how much refund we should return
+            let (currency, refund, fee, tip) = match paid {
+                PaymentInfo::Native(paid_fee) => (
+                    T::NativeAssetId::get().into(),
+                    paid_fee.saturating_sub(corrected_fee),
+                    corrected_fee.saturating_sub(tip),
+                    tip,
+                ),
+                PaymentInfo::NonNative(paid_fee, currency, price) => {
+                    // calculate corrected_fee in the non-native currency
+                    let converted_corrected_fee = price
+                        .checked_mul_int(corrected_fee)
+                        .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                    let refund = paid_fee.saturating_sub(converted_corrected_fee);
+                    let converted_tip = price
+                        .checked_mul_int(tip)
+                        .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                    (
+                        currency.into(),
+                        refund,
+                        converted_corrected_fee.saturating_sub(converted_tip),
+                        converted_tip,
+                    )
+                }
+            };
+
+            // refund to the account that paid the fees
+            MC::deposit(currency, who, refund)
+                .map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+            // deposit the fee
+            let mut fee_amounts = Vec::new();
+            if !fee.is_zero() {
+                fee_amounts.push((currency, fee));
+            }
+            if !tip.is_zero() {
+                fee_amounts.push((currency, tip));
+            }
+
+            DF::deposit_fee(&fee_receiver, fee_amounts.into_iter())
+                .map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
```

Comment:
this goes together with the other comment about deposit_fee interface.

why this needs to be done in this way ? 

what about this ?
```suggestion
           let total  = fee + tip;
           if total > 0 {
                  DF::deposit_fee(&fee_receiver, currency, total)
                      .map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
                }
```

Reply:
updated. the check for zero value is not necessary because `deposit_fee` is no-op if `total_fee == 0`